runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
+    iree_hal_hip_device_create_stream_command_buffer(
+        (iree_hal_device_t*)data->device, mode,
+        command_buffer->allowed_categories, data->queue_affinity, 0,
+        &stream_command_buffer);
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_resource_set_insert(data->resource_set, 1,
+                                         &stream_command_buffer));
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_deferred_command_buffer_apply(
+                command_buffer, stream_command_buffer, binding_table));
+    data->command_buffer = stream_command_buffer;
+  } else if (iree_hal_hip_stream_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0,
+        iree_hal_resource_set_insert(data->resource_set, 1, &command_buffer));
+  } else if (iree_hal_hip_graph_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0,
+        iree_hal_resource_set_insert(data->resource_set, 1, &command_buffer));
+    hipGraphExec_t exec =
+        iree_hal_hip_graph_command_buffer_handle(command_buffer);
+    IREE_TRACE_ZONE_BEGIN_NAMED(z3, "HipGraphLaunch");
+    status = IREE_HIP_RESULT_TO_STATUS(
+        data->device->hip_symbols,
+        hipGraphLaunch(exec,
+                       device->topology.devices[idx].hip_dispatch_stream));
+    IREE_TRACE_ZONE_END(z3);
+  } else {
+    status = iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "unsupported command buffer type");
+  }
+
+  IREE_TRACE_ZONE_END(z3);
+
+  IREE_TRACE_ZONE_BEGIN_NAMED(z4, "Record completion semaphores");
+
+  for (iree_host_size_t i = 0; i < data->signal_semaphore_list.count; ++i) {
+    iree_hal_hip_event_t* event;
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_semaphore_get_hip_event(
+                data->signal_semaphore_list.semaphores[i],
+                data->signal_semaphore_list.payload_values[i],
+                data->device->topology.devices[idx].device_event_pool, true,
+                &event));
+    if (!event) {
+      return iree_make_status(IREE_STATUS_ABORTED, "the hip event is missing");
+    }
+    status = IREE_HIP_RESULT_TO_STATUS(
+        data->device->hip_symbols,
+        hipEventRecord(iree_hal_hip_event_handle(event),
+                       device->topology.devices[idx].hip_dispatch_stream));
+    iree_hal_hip_event_release(event);
+  }
+  IREE_TRACE_ZONE_END(z4);
+
+  IREE_TRACE_ZONE_BEGIN_NAMED(z5, "Notify semaphore forward progress");
+  for (iree_host_size_t i = 0; i < data->signal_semaphore_list.count; ++i) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_semaphore_notify_forward_progress_to(
+                data->signal_semaphore_list.semaphores[i],
+                data->signal_semaphore_list.payload_values[i]));
+  }
+  IREE_TRACE_ZONE_END(z5);
+  IREE_TRACE_ZONE_BEGIN_NAMED(z6, "Record cleanup semaphore");
+
+  iree_hal_hip_event_t* event;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
```

Comment:
use vertical whitespace to separate logical blocks -> avoid vertical whitespace where it adds confusion (here keeping the event next to the thing initializing it is important to allow easier refactoring of blocks of lines without losing things)

also: always default initialize (here and elsewhere)

```suggestion
  iree_hal_hip_event_t* event = NULL;
  IREE_RETURN_AND_END_ZONE_IF_ERROR(
    ... &event);
```